### ISSUE TITLE
Router drops entire write even if only one line contains an error

### ISSUE THREAD

**Issue by alamb** [influxdata/influxdb_iox#5226]
__Steps to reproduce:__
List the minimal actions needed to reproduce the behaviour.

1. Run iox like `cargo run -vv`

2. Write this line protocol to iox (that defines `foo` to be a tag):
```
test,foo=t1 comment="foo is a tag, value t1",f=1 10000
test,foo=t2 comment="foo is a tag, value t2",f=1 10001
```
```shell
curl -v "http://127.0.0.1:8080/api/v2/write?org=00000000000000000000&bucket=917b97a92e883afc" --data-binary @example-good.lp
```

3. And then send this line protocol which has a line that now treats `foo` as a field

```
test foo="t3",comment="foo has become a field",f=1 10002
test2 comment="this is a new measurement: test2",f=1 10002
```
```shell
curl -v "http://127.0.0.1:8080/api/v2/write?org=00000000000000000000&bucket=917b97a92e883afc" --data-binary @example-bad.lp
```

Here are the files [repro.zip](https://github.com/influxdata/influxdb_iox/files/9200669/repro.zip)

__Expected behaviour:__
"Whatever cloud2 / tsm do" -- which @jacobmarble thinks is to accept the write, write the good points, and stash the old ones

So for example, if you write 1000 points and 1 is bad (mismatched tags / fields), it will accept 999 of them and ingest them and send the other to somewhere else (the `_system` bucket?)

Perhaps IOx could still respond with a 400 but write the valid points?

Or perhaps we can add a flag that toggled between "reject the entire write" or "accept write even if there is are some points that fail validation"?

__Actual behaviour:__
IOx rejects the second write request with a `400` about schema mismatch

```shell
curl -v "http://127.0.0.1:8080/api/v2/write?org=00000000000000000000&bucket=917b97a92e883afc" --data-binary @/Users/alamb/Downloads/example-bad.lp 
<.0.1:8080/api/v2/write?org=00000000000000000000&bucket=917b97a92e883afc" --data-binary @/Users/alamb/Downloads/example-bad.lp 
*   Trying 127.0.0.1:8080...
* Connected to 127.0.0.1 (127.0.0.1) port 8080 (#0)
> POST /api/v2/write?org=00000000000000000000&bucket=917b97a92e883afc HTTP/1.1
> Host: 127.0.0.1:8080
> User-Agent: curl/7.79.1
> Accept: */*
> Content-Length: 116
> Content-Type: application/x-www-form-urlencoded
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 400 Bad Request
< content-length: 140
< date: Wed, 27 Jul 2022 15:51:42 GMT
< 
* Connection #0 to host 127.0.0.1 left intact
```

Note that if you send just the second line
```
test2 comment="this is a new measurement: test2 but in a different file than bad line",f=1 10002
```
Then the write succeeds and the data is in `test2` table as expected

**Comment by mkmik** [comment]
interesting detail:

depending on the kind of schema validation error (DML error or line protocol parse error, see https://github.com/influxdata/influxdb_iox/issues/5227) the previous points may or may not be preserved:


```console
$ cat /tmp/tag.lp
bar,foo=abc v=2 1658854070000000000
$ curl -i -X POST "http://localhost:8080/api/v2/write?bucket=2b5747b96181b447&org=94af70dca3659d42&precision=ns" --data-binary @/tmp/tag.lp
HTTP/1.1 204 No Content
x-iox-write-token: eyJzZXF1ZW5jZXJzIjpbeyJzZXF1ZW5jZU51bWJlcnMiOlsiMCJdfV19
date: Wed, 27 Jul 2022 16:31:28 GMT
$ cat /tmp/field.lp
before v=1 1658854070000000000
bar foo="abc",v=2 1658854070000000000
after v=3 1658854070000000000
$ curl -i -X POST "http://localhost:8080/api/v2/write?bucket=2b5747b96181b447&org=94af70dca3659d42&precision=ns" --data-binary @/tmp/field.lp
HTTP/1.1 400 Bad Request
content-length: 140
date: Wed, 27 Jul 2022 16:32:36 GMT

{"code":"invalid","message":"dml handler error: schema conflict: column foo is type tag but write has type iox::column_type::field::string"}
$ ./target/debug/influxdb_iox query 94af70dca3659d42_2b5747b96181b447 'show tables'
+---------------+--------------------+------------+------------+
| table_catalog | table_schema       | table_name | table_type |
+---------------+--------------------+------------+------------+
| public        | iox                | bar        | BASE TABLE |
| public        | iox                | before     | BASE TABLE |
| public        | system             | queries    | BASE TABLE |
| public        | information_schema | tables     | VIEW       |
| public        | information_schema | views      | VIEW       |
| public        | information_schema | columns    | VIEW       |
+---------------+--------------------+------------+------------+
```

notice the `before` table appeared, but not `after` table.

---

Running the same test but with this file:

```console
$ cat /tmp/field.lp
before v=1 1658854070000000000
bar foo="abc",v=2 1658854070000000000
after v=3 1658854070000000000
internal,schema=validation schema="validation" 1658854070000000000
$ curl -i -X POST "http://localhost:8080/api/v2/write?bucket=2b5747b96181b447&org=94af70dca3659d42&precision=ns" --data-binary @/tmp/field.lp
HTTP/1.1 400 Bad Request
content-length: 188
date: Wed, 27 Jul 2022 16:34:19 GMT

{"code":"invalid","message":"failed to parse line protocol: error writing line 4: Unable to insert iox::column_type::field::string type into column schema with type iox::column_type::tag"}
$ ./target/debug/influxdb_iox query 94af70dca3659d42_2b5747b96181b447 'show tables'
+---------------+--------------------+------------+------------+
| table_catalog | table_schema       | table_name | table_type |
+---------------+--------------------+------------+------------+
| public        | iox                | bar        | BASE TABLE |
| public        | system             | queries    | BASE TABLE |
| public        | information_schema | tables     | VIEW       |
| public        | information_schema | views      | VIEW       |
| public        | information_schema | columns    | VIEW       |
+---------------+--------------------+------------+------------+
```

the `before` table doesn't appear because the whole batch failed to parse and thus no line was ever attempted to be written.

**Comment by philjb** [comment]
I believe we should accept as much data as we can reliably (interpret) from a batch write. The partial write of cloud2 is new-ish (early 2022, I believe). LP syntax it that the parser can't recover after encountering an error as there isn't a place in LP to jump to that is known good. 

I do wish we returned a different status code (207?) when we accept only part of a batch. The customer doesn't know to retry it with the 204 we send currently.

**Comment by mkmik** [comment]
When the error code says one line in your input data is incorrect you shouldn't retry that whole request.
Only transient errors should be retried automatically.

Imagine a customer sending a batch with 1000 lines and we return 400 because 2 lines in middle have a wrong schema or whatnot. If they retry on error, they would be effectively performing upserts for 99.8% of their ingest data. If they retry twice they will triple their write throughput and put quite some strain on compactors for no good reason.

The user should be aware that a partial write occurred, but the user should fix their data and resend the broken data or at least ensure that new occurrences of those points are correct in the future.

However, in order to do so, the user must know about all the failed lines in the batch. Currently the write response only returns an error message about the first encountered error. Returning an error message for every line in the batch can be impractical.  I think the [_monitoring bucket](https://github.com/influxdata/influxdb_iox/issues/5014) may be a better approach to handle also ingestions that happen synchronously and not only downstream of kafka. (IIRC that's what influxdb cloud2 does, all bad points are logged in the _monitoring bucket).

**Comment by philjb** [comment]
I think we are in agreement.

> Imagine a customer sending a batch with 1000 lines and we return 400 because 2 lines in middle have a wrong schema or whatnot. If they retry on error ...

Customers shouldn't be retrying 400 errors without making a modification to the request (to fix the error!).  

> However, in order to do so, the user must know about all the failed lines in the batch. 

Cloud2 does make a record in _monitoring for the first error encountered. OSS does not.

> However, in order to do so, the user must know about all the failed lines in the batch.

Yes, but depending on the error it isn't straight forward to recover LP after the error. Maybe we should have a "strict" LP subset that is strictly line based so processing can recover at the next line after an error.

**Comment by mkmik** [comment]
yeah there are LP "parse errors" which once encountered means we cannot proceed, and there are "schema validation errors" that happen after the LP has been parsed and allow us to continue ingesting other lines and potentially encountering more lines that yield a "schema validation error"

**Comment by alamb** [comment]
In past products I have worked on, we offered the user a choice of how to handle semantic errors in multi-line data as different customers wanted different things (typically depending on their upstream data system and if it was easier to remake the entire batch or if they could deal with individual rows)

One might also imagine a customer that wanted to error the entire batch as it may signal an data error in their upstream process that they wanted to fix rather than fixing up the data once it got into IOx. 

I agree there is a distinction between syntactic errors (we can't actually parse text into LP lines or we can't recover when we hit an error) and semantic ones

**Comment by mkmik** [comment]
@alamb how was the choice done? on a per API request basis? per-database/per-user configuration?

**Comment by alamb** [comment]
> @alamb how was the choice done? on a per API request basis? per-database/per-user configuration?

It was done on a per API request basis. 

https://www.vertica.com/docs/9.3.x/HTML/Content/Authoring/SQLReferenceManual/Statements/COPY/COPY.htm

So the user could do something like this which would reject the entire batch:

```sql
COPY ... FROM 'my_directory' ABORT ON ERROR
```

or insert valid records but put invalid records in some file:

```sql
COPY ... FROM 'my_directory' REJECTED DATA '/tmp/path_to_rejected_data/'
```

There is also a way to put the rejected data into another table, similar to something in the `_monitoring` bucket

**Comment by alamb** [comment]
@pauldix  described his preferred behavior here: https://github.com/influxdata/influxdb_iox/issues/5014#issuecomment-1203275533

(that is accept lines that pass, and return detailed errors for each that failed to the client)

**Comment by domodwyer** [comment]
I've been looking at _how_ this should work before looking at the code.

## Flag

> Or perhaps we can add a flag that toggled between "reject the entire write" or "accept write even if there is are some points that fail validation"?

For the sake of simplicity, I'd suggest we implement what is needed today for both practical reasons, and Cloud2 compatibility (partial writes) and defer expanding the API with a fail mode switch until it is desired by an end user. Smaller API surface == less code == less complexity == less bugs! :bug: 

## Response Format

Looking at the [InfluxDB Cloud API Documentation][cloud], under the `400 Bad Request` header (click to expand) which covers partial writes:

![image](https://user-images.githubusercontent.com/9275968/185101533-bc460826-0f4d-4e1d-8408-abcad2ee5776.png)

The response is defined as:

* **Code**: appears to be an enum of error types[^1]
* **Err**: "stack of errors" - I assume an array of error strings joined with a delimiter
* **Line**: first problematic line in the batch[^2]
* **Message**: human readable error message (not sure how it differs from `err`)
* **Op**: logical code operation that failed[^3]

Do we want to / are we required to replicate this response format? I feel we can improve it to make it more useful to the end user, as @pauldix has hinted in the above linked comment:

> Because IOx is able to validate the data and schema at the router layer, it should return an error to the client that includes any bad lines so that they can log the error and look into it on their end.


In the presence of multiple failed lines, being able to return the full set of bad lines, or their indexes, would be preferable to returning the first bad line index. Returning the full set of bad lines allows the API to fully define the "partial" nature of the write - the user can infer the full set of accepted writes, and knows the full set of rejected writes. Because the current API implementation returns a single error line index, the full set of rejected writes cannot be known by the end user without retrying N times for N bad lines.

If possible, I suggest an response such as:

```json
{
	"message": "a human readable error description",
	"line_errors": [
		{ "index": 24, "err": "human readable error"},
		{ "index": 42, "err": "another parse error"},
		{ "index": 4242, "err": "yet another big bad error"}
	]
}
```

Where the existing API field `Code` is replaced by the HTTP code and `Err` is in-lined with the actual line index that caused the error. @mkmik notes [here][mkm] that returning error strings for each line may be impractical but doesn't expand on reasoning - I therefore guesstimate it is due to the size of the response, and propose we:

* Do not store the error string N times in memory, but rather an enum/error code/int which we materialise to the string description when we write the error response to the client
* Stream the response back to the client, rather than the current [buffering response][buffering] approach taken today

That way the only cost to us that I can think of is:

* Holding the connection open longer to send the bigger response
* Network I/O & associated cost

Both of which likely have negligible impact to us, while providing a much better UX to the end user than we currently do - they can see exactly which lines are problematic (facilitating easy error discovery / retries / etc) with a descriptive reason for each rejection, in a structured manner suitable for further processing on the client's end.


[buffering]: https://github.com/influxdata/influxdb_iox/blob/180ff9f681949cc605c761dc67463fad57526448/ioxd_common/src/http/error.rs#L115-L124
[mkm]: https://github.com/influxdata/influxdb_iox/issues/5226#issuecomment-1198799958

[^1]: I'm not sure how useful this is - there don't appear to be any defined meaning of these (i.e. what's the difference / when would I expect to see "conflict" vs. "invalid" and how would I remedy the situation.)

[^2]: This doesn't define if it is 0 or 1 indexed.

[^3]: I do not know what this means :(

[cloud]:https://docs.influxdata.com/influxdb/cloud/api/#operation/PostWrite

**Comment by mkmik** [comment]
Yes. I mentioned that returning the full error message for all bad lines may be problematic in some circumstances due to response size. Not sure if that's really the case though.

**Comment by alamb** [comment]
> Do we want to / are we required to replicate this response format? I feel we can improve it to make it more useful to the end user

I am sure we can make it more useful for the user, but compatibility is a nice feature... 

Could we do both (like maintain compatibility by returning the existing fields as well as adding an extra `line_errrors` field as you proposed. Perhaps then we could say that IOx returns a compatible (though extended) response

> I therefore guesstimate it is due to the size of the response, and propose we:

Another approach that might be worth considering, if it is simpler to implement, would be to  limit the number of verbose messages returned something large but finite (like 100). The theory is if the payload has that many errors, they are likely to be mostly the same thing so additional detailed reporting is not helpful. 

Something like

```json
{
	"message": "a human readable error description",
	"line_errors": [
		{ "index": 24, "err": "human readable error"},
		{ "index": 42, "err": "another parse error"},
		{ "index": 4242, "err": "yet another big bad error"},
                .. 97 more
		{ "index": 10000 }
		{ "index": 10022 }
                 ...
	]
}
```

**Comment by domodwyer** [comment]
> Could we do both (like maintain compatibility by returning the existing fields as well as adding an extra line_errrors field as you proposed. Perhaps then we could say that IOx returns a compatible (though extended) response

I'm guessing you replied to the email and didn't see the edit I made a minute or so after the original post! At the bottom:

> Perhaps a more acceptable solution would be to add the proposed line_errors to the existing Cloud2 error response, preserving backwards compatibility but providing the benefits of the full error set.

Regarding the "first 100 errors only" it certainly is an option, but I doubt it'll be a huge difference to the network I/O / response time. If we measure that returning the verbose errors is a problem we can definitely do this, but lets avoid premature optimisation and KISS 👨‍🍳

**Comment by alamb** [comment]
> Regarding the "first 100 errors only" it certainly is an option, but I doubt it'll be a huge difference to the network I/O / response time. If we measure that returning the verbose errors is a problem we can definitely do this, but lets avoid premature optimisation and KISS 👨‍🍳

Sounds good to me. In the same vein depending on the complexity we could also avoid changing to be streaming responses as well initially

**Comment by philjb** [comment]
I recommend just returning an index of the error (not any text from the payload). The customer/client has the payload they just sent; we shouldn't repeat anything they already have in the response.

**Comment by zekena2** [comment]
Is there any updates on this? It causes dataloss as telegraf doesn't have a retrying mechanism and iox drops the valid lines + iox is more strict than influxdb v2 so things don't work right away when migrating from v2.

**Comment by philjb** [comment]
I agree as I mentioned above that ideally iox should accept lines that are fine until it can no longer unambiguously interpret more data. There hasn't been much improvement in this area.


I'm more interested in that you are saying that telegraf is producing lines that iox is rejecting - that's the bigger issue. Telegraf should only be producing line protocol that is valid for iox (or iox has a line protocol bug - wouldn't be the first).

Do you have an example of line protocol from telegraf that is being rejected? We could perhaps fix that sooner than changing iox to process as much of a batch as it can.

**Comment by zekena2** [comment]
One example that I see is when there's a schema conflict like inserting a line protocol with a tag and a field with the same name it gives an error like this.

2023-08-30T22:42:03Z E! [outputs.influxdb_v2] Failed to write metric to <bucket> (will be dropped: 400 Bad Request): invalid: failed to parse line protocol: error writing line 883: Unable to insert iox::column_type::field::integer type into column event with type iox::column_type::tag

The line I tried to import had a tag with the name `event` and a field with the same name `event`.
and AFAIK the batch gets dropped by telegraf.

**Comment by philjb** [comment]
Do you have an example (or modified failure example) for this error?

```
2023-08-30T22:51:47Z E! [outputs.influxdb_v2] Failed to write metric to bucket (will be dropped: 400 Bad Request): invalid: failed to parse line protocol: error parsing line 446 (1-based): Invalid measurement was provided
```

**Comment by zekena2** [comment]
I tried to dig deeper with one of the problematic files that I'm trying to import and found more things.

The line reports `Invalid measurement` error is because of a broken line. I'll include a file for it. but one thing I noticed is the line number reported by influxdb isn't for the actual broken line so it's misleading and makes me delete valid lines instead. I used `influxdb_iox write <namespace> filename` command with a relatively big file
[test.txt](https://github.com/influxdata/influxdb_iox/files/12484479/test.txt)

Another error is `error parsing line 641 (1-based): Could not parse entire line. Found trailing content: `. Again this error reports the wrong line number but the correct content. and it happens when the line is mixed up like this.
`h2o_temperature,location=santa_monica,state=CA surface_degrees=65.2,bottom_degrees=50.4air_temperature,location=santa_monica,state=CA sea_level_degrees=77.3,tenk_feet_feet_degrees=40.0 156875616`

in a small line protocol file iox reports the correct line number but for some reason it doesn't do that when it's a big file.

**Comment by philjb** [comment]
@zekena2 - are you using influx iox in cloud3/serverless or locally or ? 

I get that the error message line numbering is off (hard to give a good parse error really), but it seems like these are legitimately malformed line-protocol lines - right? You'd agree they should be rejected?

I am interested in cases where iox is rejecting or failing to parse a line that the oss or cloud2 golang implementation accepts. You can test against this tool: https://github.com/influxdata/line-protocol/tree/v2/cmd/lpverify. It should give better line protocol error messages. 

Fixing the issue in this ticket (accept non-failing lines in a batch) is unlikely to be worked on soon/this year (my best guess). We welcome prs though if you or your team want to pr the improvement. 

Improving the line number reporting in the errors for lp might get some traction this year (we should place it into a separate ticket).

If you have a case where the iox lp parser is failing where the go one succeeds, that's a correctness bug and we (probably me) can work on that sooner. Here's an example of that - this linked comment shows how to use the lpverify cmd i linked above: https://github.com/influxdata/influxdb_iox/pull/7997#issuecomment-1592123607

**Comment by philjb** [comment]
the `test.txt` example has a bunch of null characters at the beginning - that's a weird case - weird if telegraf passed that through without complaining. line-protocol has to be utf8 printable characters.

**Comment by philjb** [comment]
> Improving the line number reporting in the errors for lp might get some traction this year (we should place it into a separate ticket).

If you have an example that reproduces the bad line numbers in error messages - we can create a separate ticket for that. I couldn't recreate it from your samples (and some guessing at what to do).

**Comment by zekena2** [comment]
I'm using iox locally but the same things appear when I try the cloud version.

I agree that influxdb shouldn't accept these lines but as I said it's simply harder to import the same stuff I used to import in v2 because of mainly rejecting good lines and the wrong line number makes it even harder and more data gets lost.
I'm fine with waiting until it gets fixed. I'd like to contribute but I'm not good with rust tbh. 
Thanks for the link for the lp verifier I was actually searching for something like this in order to catch these lines before sending them to iox.

I'll try to make a reproducible example of wrong line numbers in responses I only found out about it today.

**Comment by philjb** [comment]
@zekena2 - glad the lp verifier is helpful for you. It'd be nice if it was better known. Thanks for replying about cloud/local and for keeping your eye out for an example with the error reporting the wrong line (I remember the golang version struggling with this at one point).

@wiedld is working on incremental improvements for this ticket - This makes it sound like I asked her - I didn't. You should get something sooner than before now that she is looking at it and before no one was! 

Thanks for being active here.

**Comment by wiedld** [comment]
I have been breaking this ticket up into 2 main goals, and a question was raised this morning of whether or not the first goal should be done. Parts:
* enable itemized line errors in either mode (partial writes, or all-or-none) up to 100 errors.
* provide a config (per request) to do partial writes versus all-or-none.

What are people's thoughts on the first item?

**Comment by pauldix** [comment]
Why the limit to 100 errors? Why not just return an error for every line that has one?

**Comment by alamb** [comment]
> Why the limit to 100 errors? Why not just return an error for every line that has one?

@pauldix  -- I believe the rationale is to avoid pathological cases consuming too much bandwidth / CPU on the server or networks. For example, if some formatting thing messed up all lines for 10MB of line protocol, the resulting error messages for all 10MB could be substantially more (like 100s of MB of error messages)

Also, it seems unlikely that someone will manually review / handle 100 errors (let alone thousands) so the extra value gained from so many verbose error messages is low.  

If we bound the number of detailed error messages we prevent the pathological case. 


https://github.com/influxdata/influxdb_iox/issues/5226#issuecomment-1218123852

**Comment by wiedld** [comment]
The specific question brought up this morning -- is whether the itemized errors (potentially bounded) should be returned for only the partial writes? or also for the current behavior (reject on any error)?

Currently, when we reject on any error, we only return the first error encountered. We do not return a list of errors -- therefore the user has to work through each line error one-by-one.

The [proposed change here](https://github.com/influxdata/influxdb_iox/pull/8690) would return a list of errors in either case, regardless of whether or not there is partial writes configured.

**Comment by jacobmarble** [comment]
I'm bounding the requirements of this feature.

# Goals

- Support InfluxDB 1.x and 2.x clients.
- Do not add to [InfluxDB 1.x and 2.x APIs](https://docs.influxdata.com/influxdb/cloud-serverless/api/#operation/PostWrite).
- Do not (or gracefully) break existing IOx clients.
- Do not fail entire line protocol payloads; only fail offending lines.

# Partial Error Response

Partial failures return 400. The response body looks like:
```json
{
  "code": "invalid",
  "line": 123,
  "message": "human-readable error message"
}
```

`code` is `"invalid"`.

`line` is the first failing line.

`message` may contain details of up to 100 failing lines. This is a human-readable value, so we can't break the 1.x/2.x API this way.

# Modes

We need to _temporarily_ support current and improved behavior, hence two "modes". The mode is configured via process flag and/or env var. In either mode, a given failure returns the same human-readable information.

**All-Or-None**: Current behavior; entire payloads are rejected.

**Partial-Error**: New behavior; non-failing lines are applied.

# Plan to Remove Modes

When the change is merged, the default mode is All-Or-None. This way, service owners may choose to test Partial-Error mode.

About two weeks after the change is merged, Partial-Error mode becomes the hard-coded behavior, and the temporary "modes" config is removed.

# `_monitoring` bucket

This would be nice to have, but let's plan it separate from this issue.
